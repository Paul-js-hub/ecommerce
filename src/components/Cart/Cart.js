import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from 'react-router-dom';
import { 
  getCartCount, 
  getCartProducts, 
  getSubTotal,
  removeCartItem, 
  increment, 
  decrement 
} from "../../redux/cart/cart";
import {
  MDBBtn,
  MDBCard,
  MDBCardBody,
  MDBCardImage,
  MDBCol,
  MDBContainer,
  MDBIcon,
  MDBRow,
  MDBCardText,
  MDBTypography,
} from "mdb-react-ui-kit";
import "./cart.css";

const Cart = () => {
  const dispatch = useDispatch();
  const { cart, totalCount } = useSelector((state) => state.cart);
  let total = 0;
  cart.map(product => {
    return (total = total + product.price * product.quantity)
  })
  useEffect(() => {
    dispatch(getCartProducts);
    dispatch(getCartCount())
    dispatch(getSubTotal());
  }, [dispatch]);

  let displayCart;
  if (cart !== undefined && cart.length > 0) {
    displayCart = (
      <>
        <section className="h-100 h-custom" style={{ backgroundColor: "#eee" }}>
          <MDBContainer className="py-5 h-100">
            <MDBRow className="justify-content-center align-items-center h-100">
              <MDBCol size="12">
                <MDBCard
                  className="card-registration card-registration-2"
                  style={{ borderRadius: "15px" }}
                >
                  <MDBCardBody className="p-0">
                    <MDBRow className="g-0">
                      <MDBCol lg="8">
                        <div className="p-5">
                          <div className="d-flex justify-content-between align-items-center mb-5">
                            <MDBTypography
                              tag="h1"
                              className="fw-bold mb-0 text-black"
                            >
                              Shopping Cart
                            </MDBTypography>
                            <MDBTypography className="mb-0 text-muted">
                            {totalCount} item(s) added
                            </MDBTypography>
                          </div>

                          <hr className="my-4" />

                          {cart !== undefined &&
                            cart.length > 0 &&
                            cart.map((product, index) => {
                              return (
                                <MDBRow className="mb-4 d-flex justify-content-between align-items-center" key={index}>
                                  <MDBCol md="2" lg="2" xl="2">
                                    <MDBCardImage
                                      src={product.image}
                                      fluid
                                      className="rounded-3"
                                      alt="product"
                                    />
                                  </MDBCol>
                                  <MDBCol md="3" lg="3" xl="3">
                                    <MDBTypography
                                      tag="h6"
                                      className="text-black"
                                    >
                                      {product.title}
                                    </MDBTypography>
                                    <MDBTypography
                                      tag="h6"
                                      className="text-muted mb-0"
                                    >
                                      {product.description.substring(0, 21)}...
                                    </MDBTypography>
                                  </MDBCol>
                                  <MDBCol
                                    md="3"
                                    lg="3"
                                    xl="3"
                                    className="d-flex align-items-center"
                                  >
                                    <MDBBtn color="link" className="px-2">
                                      <MDBIcon fas icon="minus" 
                                       onClick={() => {
                                        dispatch(decrement(product.id))
                                        dispatch(getSubTotal())
                                        dispatch(getCartCount())
                                      }}
                                      />
                                    </MDBBtn>

                                    <MDBTypography
                                    tag="h6"
                                    className="text-muted mb-0"
                                    >
                                      {product.quantity}
                                    </MDBTypography>

                                    <MDBBtn color="link" className="px-2">
                                      <MDBIcon fas icon="plus"
                                      onClick={() => {
                                        dispatch(increment(product.id))
                                        dispatch(getSubTotal())
                                        dispatch(getCartCount())
                                      }}
                                      />
                                    </MDBBtn>
                                  </MDBCol>
                                  <MDBCol
                                    md="3"
                                    lg="2"
                                    xl="2"
                                    className="text-end"
                                  >
                                    <MDBTypography tag="h6" className="mb-0">
                                      {product.price}
                                    </MDBTypography>
                                  </MDBCol>
                                  <MDBCol
                                    md="1"
                                    lg="1"
                                    xl="1"
                                    className="text-end"
                                  >
                                    <a href="#!" className="text-muted">
                                      <MDBIcon fas icon="times" 
                                      onClick={() => {
                                        dispatch(removeCartItem(product.id))
                                        dispatch(getSubTotal())
                                        dispatch(getCartCount())
                                      }}
                                      />
                                    </a>
                                  </MDBCol>
                                </MDBRow>
                              );
                            })}
                          <hr className="my-4" />

                          <div className="pt-5">
                            <MDBTypography tag="h6" className="mb-0">
                              <MDBCardText
                                tag="a"
                                href="#!"
                                className="text-body"
                              >
                                <MDBIcon fas icon="long-arrow-alt-left me-2" />{" "}
                                <Link to="/">Back to shop</Link>
                              </MDBCardText>
                            </MDBTypography>
                          </div>
                        </div>
                      </MDBCol>
                      <MDBCol lg="4" className="bg-grey">
                        <div className="p-5">
                          <MDBTypography
                            tag="h3"
                            className="fw-bold mb-5 mt-2 pt-1"
                          >
                            Summary
                          </MDBTypography>

                          <hr className="my-4" />

                          <div className="d-flex justify-content-between mb-4">
                            <MDBTypography tag="h5" className="text-uppercase">
                            {totalCount} item(s) added
                            </MDBTypography>
                            <MDBTypography tag="h5">$ {total}</MDBTypography>
                          </div>

                          <hr className="my-4" />

                          <div className="d-flex justify-content-between mb-5">
                            <MDBTypography data-testid="total-text" tag="h5" className="text-uppercase">
                              Total price
                            </MDBTypography>
                            <MDBTypography tag="h5">$ {total}</MDBTypography>
                          </div>
                        </div>
                      </MDBCol>
                    </MDBRow>
                  </MDBCardBody>
                </MDBCard>
              </MDBCol>
            </MDBRow>
          </MDBContainer>
        </section>
      </>
    );
  } else {
    displayCart = (
      <div className="w-100 alert alert-warning mt-5">
        <h2 className="h4 text-center mb-3">Your cart is empty</h2>
        <p className="text-center">Looks like you have not added products to your cart. Let's buy some.</p>
        <div className="text-center">
          <Link to="/" className="btn btn-danger">Shop now</Link>
        </div>
      </div>
    );
  }

  return <>{displayCart}</>;
};

export default Cart;
