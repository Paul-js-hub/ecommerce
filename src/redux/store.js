import { configureStore } from '@reduxjs/toolkit';
import productsReducer from './products/products.js';
import cartsReducer from './cart/cart.js';

const store = configureStore({
  reducer: {
    products: productsReducer,
    cart: cartsReducer
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware({ serializableCheck: false})
});

export default store;
