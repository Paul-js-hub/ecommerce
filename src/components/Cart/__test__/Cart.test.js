import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { HashRouter as Router } from "react-router-dom";
import store from "../../../redux/store";
import Cart from "../Cart";

describe("Test Products Component", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    )
    .toJSON();
  it("renders cart items", () => {
    expect(tree).toMatchSnapshot();
  });

  it("displays total price text", () => {
    const tree = render(
      <Provider store={store}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    );
    const element = screen.getByRole("heading", {
      name: "Your cart is empty",
    });
    expect(element).toBeInTheDocument();
  });
});
