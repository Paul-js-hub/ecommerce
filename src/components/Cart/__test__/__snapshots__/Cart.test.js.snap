// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Products Component renders cart items 1`] = `
<div
  className="w-100 alert alert-warning mt-5"
>
  <h2
    className="h4 text-center mb-3"
  >
    Your cart is empty
  </h2>
  <p
    className="text-center"
  >
    Looks like you have not added products to your cart. Let's buy some.
  </p>
  <div
    className="text-center"
  >
    <a
      className="btn btn-danger"
      href="#/"
      onClick={[Function]}
    >
      Shop now
    </a>
  </div>
</div>
`;
